src/spac/data_utils.py

```diff
@@ -936,3 +937,127 @@ def combine_dfs(dataframes: list):
     combined_df.reset_index(drop=True, inplace=True)
 
     return combined_df
+
+
+def add_pin_color_rules(
+    adata,
+    label_color_dict: dict,
+    color_map_name: str,
+    overwrite: bool = True
+) -> Tuple[dict, str]:
+    """
+    Adds pin color rules to the AnnData object and scans for matching labels.
+
+    This function scans unique labels in each adata.obs and column names in all
+    adata tables, to find the labels defined by the pin color rule.
+
+    Parameters
+    ----------
+    adata
+        The anndata object containing upstream analysis.
+    label_color_dict : dict
+        Dictionary of pin color rules with label as key and color as value.
+    color_map_name : str
+        The name to use for storing pin color rules in `adata.uns`.
+    overwrite : bool, optional
+        Whether to overwrite existing pin color rules in `adata.uns` with the
+        same name, by default True.
+
+    Returns
+    -------
+    label_matches : dict
+        Dictionary with the matching labels in each
+        section (obs, var, X, etc.).
+    result_str : str
+        Summary string with the matching labels in each
+        section (obs, var, X, etc.).
+
+    Raises
+    ------
+    ValueError
+        If `color_map_name` already exists in `adata.uns`
+        and `overwrite` is False.
+    """
+
+    # Check if the pin color rule already exists in adata.uns
+    if color_map_name in adata.uns and not overwrite:
+        raise ValueError(
+            f"`{color_map_name}` already exists in `adata.uns` ",
+            "and `overwrite` is set to False."
+        )
+
+    # Add or overwrite pin color rules in adata.uns
+    adata.uns[color_map_name] = label_color_dict

```

Comment:
@ruiheesi, so you save the dictionary even before looking for the labels? If that's the case, the purpose of the rest of the function is to generate the summary of where the labels are? 

If so, should we save that summary in the anndata object too?

Reply:
Hi @georgezakinih, the pin color step can happen anywhere in the analysis, it is not a required step.

For example, say in a consistent set of work we always want PIMO+ label as red, before we actually run any analysis, we can add PIMO+: red into the anndata object for future usage. While doing this, the check returns the warning saying PIMO+ does not exists. Another scenario is when we done with analysis and we want to go for viz, we want to keep PIMO+ red, and adding the dictionary into the analysis would help us check if PIMO+ label had been defined. 
This is a curtesy check for users rather than a hard check for addition.  

We can certainly add the log into the analysis as well. Do you prefer me to add this? Thanks.

Reply:
Hi @ruiheesi, yes please add the summary in the uns part of the anndata object.